numpy/random/mt19937.pyx

```diff
@@ -0,0 +1,364 @@
+        sequence) of such integers, or ``None`` (the default).  If `seed` is
+        ``None``, then will attempt to read data from ``/dev/urandom``
+        (or the Windows analog) if available or seed from the clock otherwise.
+
+    Notes
+    -----
+    ``MT19937`` directly provides generators for doubles, and unsigned 32 and 64-
+    bit integers [1]_ . These are not directly available and must be consumed
+    via a ``RandomGenerator`` object.
+
+    The Python stdlib module "random" also contains a Mersenne Twister
+    pseudo-random number generator.
+
+    **State and Seeding**
+
+    The ``MT19937`` state vector consists of a 768 element array of
+    32-bit unsigned integers plus a single integer value between 0 and 768
+    indicating  the current position within the main array.
+
+    ``MT19937`` is seeded using either a single 32-bit unsigned integer
+    or a vector of 32-bit unsigned integers.  In either case, the input seed is
+    used as an input (or inputs) for a hashing function, and the output of the
+    hashing function is used as the initial state. Using a single 32-bit value
+    for the seed can only initialize a small range of the possible initial
+    state values.
+
+    **Compatibility Guarantee**
+
+    ``MT19937`` make a compatibility guarantee. A fixed seed and a fixed
+    series of calls to ``MT19937`` methods will always produce the same
+    results up to roundoff error except when the values were incorrect.
+    Incorrect values will be fixed and the version in which the fix was
+    made will be noted in the relevant docstring.
+
+    **Parallel Features**
+
+    ``MT19937`` can be used in parallel applications by
+    calling the method ``jump`` which advances the state as-if :math:`2^{128}`
+    random numbers have been generated ([1]_, [2]_). This allows the original sequence to
+    be split so that distinct segments can be used in each worker process.  All
+    generators should be initialized with the same seed to ensure that the
+    segments come from the same sequence.
+
+    >>> from numpy.random.entropy import random_entropy
+    >>> from numpy.random import RandomGenerator, MT19937
+    >>> seed = random_entropy()
+    >>> rs = [RandomGenerator(MT19937(seed)) for _ in range(10)]
+    # Advance each MT19937 instance by i jumps
+    >>> for i in range(10):
+    ...     rs[i].brng.jump(i)
+
+    References
+    ----------
+    .. [1] Hiroshi Haramoto, Makoto Matsumoto, and Pierre L\'Ecuyer, "A Fast
+        Jump Ahead Algorithm for Linear Recurrences in a Polynomial Space",
+        Sequences and Their Applications - SETA, 290--298, 2008.
+    .. [2] Hiroshi Haramoto, Makoto Matsumoto, Takuji Nishimura, François
+        Panneton, Pierre L\'Ecuyer, "Efficient Jump Ahead for F2-Linear
+        Random Number Generators", INFORMS JOURNAL ON COMPUTING, Vol. 20,
+        No. 3, Summer 2008, pp. 385-390.
+
+    """
+    cdef mt19937_state *rng_state
+    cdef brng_t *_brng
```

Comment:
Is there a reason this is on the heap and not just within the object itself? The indirection here doesn't seem useful to me.

Reply:
Probably no good reason other than looking like the line above. These can be changed.  the references to brng_t should probably also be replaced with something like bitgen_t given the decisions taken about names.